okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -993,33 +1029,36 @@ private void resolvePath(String input, int pos, int limit) {
     /** Adds a path segment. If the input is ".." or equivalent, this pops a path segment. */
     private void push(String input, int pos, int limit, boolean addTrailingSlash,
         boolean alreadyEncoded) {
-      int segmentLength = limit - pos;
-      if ((segmentLength == 2 && input.regionMatches(false, pos, "..", 0, 2))
-          || (segmentLength == 4 && input.regionMatches(true, pos, "%2e.", 0, 4))
-          || (segmentLength == 4 && input.regionMatches(true, pos, ".%2e", 0, 4))
-          || (segmentLength == 6 && input.regionMatches(true, pos, "%2e%2e", 0, 6))) {
+      String segment = canonicalize(
+          input, pos, limit, PATH_SEGMENT_ENCODE_SET, alreadyEncoded, false);
+      if (isDot(segment)) {
+        return; // Skip '.' path segments.
+      }
+      if (isDotDot(segment)) {
         pop();
         return;
       }
-
-      if ((segmentLength == 1 && input.regionMatches(false, pos, ".", 0, 1))
-          || (segmentLength == 3 && input.regionMatches(true, pos, "%2e", 0, 3))) {
-        return; // Skip '.' path segments.
-      }
-
-      String segment = canonicalize(
-          input, pos, limit, PATH_SEGMENT_ENCODE_SET, alreadyEncoded, false);
       if (encodedPathSegments.get(encodedPathSegments.size() - 1).isEmpty()) {
         encodedPathSegments.set(encodedPathSegments.size() - 1, segment);
       } else {
         encodedPathSegments.add(segment);
       }
-
       if (addTrailingSlash) {
         encodedPathSegments.add("");
       }
     }
 
+    private boolean isDot(String input) {
+      return input.equalsIgnoreCase(".") || input.equalsIgnoreCase("%2e");
```

Comment:
Seemed simpler to keep all cases symmetric. Probably unnecessary. I eyeballed it both ways.

I'll change it.